### ISSUE TITLE
Optimize compilation speed and binary size

### ISSUE THREAD

**Issue by Daandelange** [d3cod3/Mosaic-Plugin#2]
Everything works great, compilation is smooth and works out-of-the-box.
But now I have a `Number Object` that takes as much time to compile as Mosaic does, plus it's the same size. (+50Mb !!) Apparently, all symbols are duplicated into the plugin binary.
Did I miss something ?

What are your findings and explorations around this ? I guess there's a reason for this _[temporary?]_ choice ?
I just checked the recently new [ofxHotReloader](https://github.com/gumilastik/ofxHotReloader) and the plugin is a little over 10Mb, which is about the size of OF, so it probably has the same approach.

Theoretically, there should be a way to simply compile the plugin code in a few bytes, then link that against the symbols included in the Mosaic binary ?
Maybe Mosaic should link with a stand-alone OpenFrameworks.a (.bundle?), then the plugins do the same ?

**Comment by d3cod3** [comment]
Yes, everything you're saying is a temporary situation, i had priority in make it work smoothly and easy, now the next step will be optimize, in order to reduce plugin size and compile speed.

Basically, we'll need to reorganize the code in order to have the minimum includes as possible to compile an object ( a plugin in this case ), then everything will be lighter and faster to compile.

Right now, every objects depends on ```PatchObject``` class

that needs:

ofxImGui
ofxFontStash
ofxPDSP
ofxXmlSettings

and most of the core/ files in ofxVisualProgramming

then, starting by creating a specific version of ofxVisualProgramming ( maybe call it ofxMosaicPlugin? ) just for plugin compiling, we'll reduce a lot the plugin size and the compile time.

ofxHotReloader is based on the older ofxPlugin, both tested a while ago, but  the smoother and bugless option was the one i used with PUGG, so i stayed on that, but yes, the principle is similar, the plugin carry a lot of unnecessary symbols at runtime, but necessary for compile the dynamic lib.

More optimizations will be possible too, as you suggest, so let's start digging the issue!

**Comment by Daandelange** [comment]
Ok, good to hear.
So there seem to be 2 aspects on this issue :
 1. Reduce the plugin includes & dependencies (easy)
 2. Figure out how to make Mosaic and Mosaic-Plugin against the same OF library (and other shared ones?).
    2.a Compile OF as a (dynamic/static?) shared library, link both against it.
    2.b Link the plugins against the OF contained within Mosaic ?

This will probably also bring changes to the OF API mappings / bridging. (maybe solving the duplicate singletons issue with the renderers ?)

Anyways, there seem to be some half-answered topics around this. [ [-1-](https://forum.openframeworks.cc/t/can-i-change-empty-example-xcode-project-output-to-dylib/25456) / [-2-](https://forum.openframeworks.cc/t/build-of-as-a-shared-library-in-macos/34132) / [-3-](https://forum.openframeworks.cc/t/build-of-as-a-dll-under-windows/19019) / [-4-](https://gist.github.com/ofZach/e602efb4fbd43858aa01aff6d015ef32) / [-5-](https://github.com/mazbox/livecode-cplusplus) ]
- [Some interesting notes on different approaches](https://github.com/caiorss/C-Cpp-Notes/blob/master/DLL-Binary-Components-SharedLibraries.org)
- [An alternative to Pugg](https://github.com/macmade/dlib/tree/master/Unit-Tests).
- [Linking the DLL against  Mosaic's embedded OF is possible ?](https://forum.openframeworks.cc/t/can-openframeworks-do-live-coding-and-how/19322/49)

Edit:
I've been digging this a bit. This is gonna get complicated, leaving this for later. The solution mentioned in the last link (above) looks the best, but I haven't succeeded replicating it with Qt.
Recommendations and [existing implementations](https://github.com/pure-data/externals-howto/blob/master/example1/helloworld.c) seem to restrict plugin compilation to use only shared dylibs, when code is shared between host and plugins. That would mean : make ImGui Shared, make OF shared, make (parts of?) ofxVP shared, etc.

**Comment by d3cod3** [comment]
Fixed point 1!

https://github.com/d3cod3/ofxMosaicPlugin

compile time drastically reduced and binary size at ~= 27% of the previous size, not bad

**Comment by Daandelange** [comment]
For point 2, I think it's worth waiting for OF+CMake support which, I think, will greatly facilitate cross linking issues.
